File: writing-core/src/main/java/dev/bee/kanjianki/core/study/WritingHintPolicy.java
package dev.bee.kanjianki.core.study;

public final class WritingHintPolicy {
    private WritingHintPolicy() {
    }

    public static HintState initialHintState(
            int writingLevel,
            int totalReviews,
            int learningStep,
            boolean targetedWriting
    ) {
        int stored = Math.max(HintLevel.TRACE.writingLevel(), Math.min(HintLevel.BLIND.writingLevel(), writingLevel));
        if (targetedWriting || totalReviews == 0 || learningStep == 0) {
            return HintState.fromWritingLevel(Math.min(stored, HintLevel.OUTLINE.writingLevel()));
        }
        return HintState.fromWritingLevel(stored);
    }
}

File: writing-core/src/main/kotlin/dev/bee/kanjianki/core/study/WritingHintPolicy.kt
package dev.bee.kanjianki.core.study

class WritingHintPolicy private constructor() {
    companion object {
        @JvmStatic
        fun initialHintState(
            writingLevel: Int,
            totalReviews: Int,
            learningStep: Int,
            targetedWriting: Boolean,
        ): HintState {
            val stored = maxOf(HintLevel.TRACE.writingLevel(), minOf(HintLevel.BLIND.writingLevel(), writingLevel))
            if (targetedWriting || totalReviews == 0 || learningStep == 0) {
                return HintState.fromWritingLevel(minOf(stored, HintLevel.OUTLINE.writingLevel()))
            }
            return HintState.fromWritingLevel(stored)
        }
    }
}
